src/lib/operators/projection.cpp

```diff
@@ -97,47 +97,109 @@ std::shared_ptr<const Table> Projection::_on_execute() {
+  auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+  jobs.reserve(chunk_count);
 
-  const auto chunk_count = input_table.chunk_count();
+  const auto expression_count = expressions.size();
+
+  // Create a two-dimensional vector that saves the information if a column is nullable for every segment.
+  // This allows parallel write operation per thread. Later this data structure is aggregated
+  // to a vector with information if a column is nullable for every column.
+  auto column_is_nullable_by_chunk = std::vector<std::vector<bool>>(
+    chunk_count,
+    std::vector<bool>(expression_count, false));
+
+  // All projections that operate on chunks of sufficient size and newly generated columns will be executed after the
+  // loop by the job scheduler to parallelize the evaluation operation.
   for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
     const auto input_chunk = input_table.get_chunk(chunk_id);
     Assert(input_chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
 
-    auto output_segments = Segments{expressions.size()};
+    auto output_segments = Segments{expression_count};
+    auto forwared_count = size_t{0};
 
-    // The ExpressionEvaluator is created once per chunk so that evaluated sub-expressions can be reused across columns.
-    ExpressionEvaluator evaluator(left_input_table(), chunk_id, uncorrelated_subquery_results);
-
-    for (auto column_id = ColumnID{0}; column_id < expressions.size(); ++column_id) {
+    for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
+      // In this loop, we perform all projections that only forward an input column sequential.
       const auto& expression = expressions[column_id];
-
       if (expression->type == ExpressionType::PQPColumn) {
         // Forward input column if possible
+        forwared_count++;
         const auto& pqp_column_expression = static_cast<const PQPColumnExpression&>(*expression);
         output_segments[column_id] = input_chunk->get_segment(pqp_column_expression.column_id);
-        column_is_nullable[column_id] =
-            column_is_nullable[column_id] || input_table.column_is_nullable(pqp_column_expression.column_id);
-        forwarding_cost += timer.lap();
-      } else {
-        // Newly generated column - the expression needs to be evaluated
-        auto output_segment = evaluator.evaluate_expression_to_segment(*expression);
-        column_is_nullable[column_id] = column_is_nullable[column_id] || output_segment->is_nullable();
-
-        // Storing the result in output_segments means that the vector may contain both ReferenceSegments and
-        // ValueSegments. We deal with this later.
-        output_segments[column_id] = std::move(output_segment);
-        expression_evaluator_cost += timer.lap();
+        column_is_nullable_by_chunk[chunk_id][column_id] =
+          column_is_nullable_by_chunk[chunk_id][column_id] ||
```

Comment:
I must be missing something. When would `column_is_nullable_by_chunk[chunk_id][column_id]` hold anything other than the default `false` falue from above?

Reply:
Good point. I think we don't need the logical `or` operator, since we are only accessing `column_is_nullable_by_chunk[chunk_id][column_id]` once.

Reply:
I change the code. We are now using a one-dimensional vector `column_is_nullable` with `atomic bool` values instead of  `column_is_nullable_by_chunk` . But I think the logical OR operation per chunk makes not so much sense (I think). One solution could be  to move this part (column_is_nullable[column_id] = ...) at the beginning of the loop and set the value with `input_table.column_is_nullable(pqp_column_expression.column_id)` for every column. Do you think that would be a good solution?

Reply:
Ah. Good point. Yes, for forwarded columns, it does indeed not make sense to do this in the loop and we can simply use the information from the input table. For newly generated columns, we still need to collect this information from the ExpressionEvaluator.